subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy

```diff
@@ -678,6 +686,52 @@ class CachedCustomTaskExecutionIntegrationTest extends AbstractIntegrationSpec i
         expectedOutput.file
     }
 
+    @Unroll
+    def "local state declared via #api API is destroyed when task is loaded from cache"() {
+        def localStateFile = file("local-state.json")
+        buildFile << defineTaskWithLocalState(useRuntimeApi)
+
+        when:
+        withBuildCache().succeeds "customTask"
+        then:
+        executedAndNotSkipped ":customTask"
+        localStateFile.assertIsFile()
+
+        when:
+        cleanBuildDir()
+        withBuildCache().succeeds "customTask"
+        then:
+        skipped ":customTask"
+        localStateFile.assertDoesNotExist()
+
+        where:
+        useRuntimeApi << [true, false]
+        api = useRuntimeApi ? "runtime" : "annotation"
+    }
+
+    @Unroll
+    def "local state declared via #api API is not destroyed when task is not loaded from cache"() {
+        def localStateFile = file("local-state.json")
+        buildFile << defineTaskWithLocalState(useRuntimeApi)
+
+        when:
+        succeeds "customTask"
+        then:
+        executedAndNotSkipped ":customTask"
+        localStateFile.assertIsFile()
+
+        when:
+        cleanBuildDir()
+        succeeds "customTask"
+        then:
+        executedAndNotSkipped ":customTask"
+        localStateFile.assertIsFile()
```

Comment:
Wouldn't we need to check in the task that the `localStateFile` exists? If it would have been deleted the task would recreate it anyway.

Reply:
Good point, I made it better.